From 1407b9b035d7d292873641fea21d71406cb38472 Mon Sep 17 00:00:00 2001
From: Joshua Berne <berne@notadragon.com>
Date: Thu, 17 Sep 2026 13:40:42 -0400
Subject: [PATCH] c++/contracts: assertion predicate re-invoking constexpr
 function [PR125459]

The RESULT_DECL of a function during constant evaluation has its value stored
in the constexpr_global_ctx, and once its lifetime ends void_list_node is put
in its place.  If the same function is then called later during the same
constant evaluation from within the scope of a modifiable_tracker (i.e.,
during speculative evaluation of a contract assertion, or evaluation of the
predicate of an [[assume]]) the population of the return value will find an
entry (void_list_node) already in the context and treat the call as an
attempt to modify outside state, resulting in errors.

This change updates the put_value member of constexpr_global_ctx to recognize
void_list_node values in the map as not being a concern for what modifiables
is tracking, and treats them the same way it treats not yet having a value in
the map for that key.  void_node is treated the same way if it is found in
the map to be consistent with other handling of the values that are stored,
though no case seems to currently hit that path.

PR c++/125459 - [contracts] "contract condition is not constant" in constexpr function

	PR c++/125459

gcc/cp/ChangeLog:

	* constexpr.cc (constexpr_global_ctx::put_value): Treat void_node
	and void_list_node as not having been in the map for modifiable
	tracking.

gcc/testsuite/ChangeLog:

	* g++.dg/contracts/cpp26/pr125459.C: Test repeated calls to constexpr
	function from within a contract assertion.

Signed-off-by: Joshua Berne <berne@notadragon.com>
---
 gcc/cp/constexpr.cc                             |  9 +++++++--
 gcc/testsuite/g++.dg/contracts/cpp26/pr125459.C | 13 +++++++++++++
 2 files changed, 20 insertions(+), 2 deletions(-)
 create mode 100644 gcc/testsuite/g++.dg/contracts/cpp26/pr125459.C

diff --git a/gcc/cp/constexpr.cc b/gcc/cp/constexpr.cc
index 18d361c2432..a6969c96f63 100644
--- a/gcc/cp/constexpr.cc
+++ b/gcc/cp/constexpr.cc
@@ -1276,8 +1276,13 @@ public:
   }
   void put_value (tree t, tree v)
   {
-    bool already_in_map = values.put (t, v);
-    if (!already_in_map && modifiable)
+    bool already_in_map;
+    tree& slot = values.get_or_insert (t, &already_in_map);
+    bool newval = !already_in_map
+		  || slot == void_node
+		  || slot == void_list_node;
+    slot = v;
+    if (newval && modifiable)
       modifiable->add (t);
   }
   void destroy_value (tree t, bool past_storage_end = true)
diff --git a/gcc/testsuite/g++.dg/contracts/cpp26/pr125459.C b/gcc/testsuite/g++.dg/contracts/cpp26/pr125459.C
new file mode 100644
index 00000000000..83f312e1611
--- /dev/null
+++ b/gcc/testsuite/g++.dg/contracts/cpp26/pr125459.C
@@ -0,0 +1,13 @@
+// { dg-do compile { target c++26 } }
+// { dg-additional-options "-fcontracts -fcontract-evaluation-semantic=enforce" }
+
+constexpr bool f(int) { return true; }
+
+constexpr bool g()
+{
+    f(0);
+    contract_assert(f(1));  // force another evaluation of f
+    return true;
+}
+
+static_assert(g());
-- 
2.43.0

